rust-runtime/aws-smithy-http-server/src/plugin/pipeline.rs

```diff
@@ -0,0 +1,143 @@
+/// so far - e.g. you only want to apply those plugins for a specified operation.
+/// You can use the [`map`](PluginPipeline::map) method to grab the current pipeline and transform it:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{filter_by_operation_name, PluginPipeline};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+/// # struct CheckHealth;
+/// # impl CheckHealth { const NAME: &'static str = "MyName"; }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     .push(MetricsPlugin)
+///     .map(|current_pipeline| {
+///         // The logging and metrics plugins will not be applied to the `CheckHealth` operation.
+///         filter_by_operation_name(current_pipeline, |name| name != CheckHealth::NAME)
+///     })
+///     // The auth plugin will be applied to all operations
+///     .push(AuthPlugin);
+/// ```
+///
+/// ## Concatenating two plugin pipelines
+///
+/// `PluginPipeline` is a good way to bundle together multiple plugins, ensuring they are all
+/// registered in the correct order.
+///
+/// You can use the [`concat`](PluginPipeline::concat) to append, at once, all the plugins
+/// in another pipeline to the current pipeline:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as MetricsPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub fn get_bundled_pipeline() -> PluginPipeline<PluginStack<LoggingPlugin, MetricsPlugin>> {
+///     PluginPipeline::new(LoggingPlugin).push(MetricsPlugin)
+/// }
+///
+/// let pipeline = PluginPipeline::new(AuthPlugin)
+///     .concat(get_bundled_pipeline());
+/// ```
+///
+/// ## Providing custom methods on `PluginPipeline`
+///
+/// You use an **extension trait** to add custom methods on `PluginPipeline`.
+///
+/// This is a simple example using `AuthPlugin`:
+///
+/// ```rust
+/// use aws_smithy_http_server::plugin::{PluginPipeline, PluginStack};
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as LoggingPlugin;
+/// # use aws_smithy_http_server::plugin::IdentityPlugin as AuthPlugin;
+///
+/// pub trait AuthPluginExt<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>>;
+/// }
+///
+/// impl<CurrentPlugins> AuthPluginExt<CurrentPlugins> for PluginPipeline<CurrentPlugins> {
+///     fn with_auth(self) -> PluginPipeline<PluginStack<CurrentPlugins, AuthPlugin>> {
+///         self.push(AuthPlugin)
+///     }
+/// }
+///
+/// let pipeline = PluginPipeline::new(LoggingPlugin)
+///     // Our custom method!
+///     .with_auth();
+/// ```
+pub struct PluginPipeline<P>(P);
```

Comment:
I'm unsure whether `PluginPipeline` is a good name. A pipeline conveys that flow can be forked and merged, whereas here we're just layering plugins. I thought our intention was for `PluginStack` to mirror `tower::layer::util::Stack`, and `PluginBuilder` to mirror `tower::ServiceBuilder`.